swift/README.md

```diff
@@ -1,372 +1,1100 @@
 
 ```swift
-let regex = NSRegularExpression(pattern: "[a-z]", options: NSRegularExpressionOptions.allZeros, error: nil)!
+func dateFromString(dateString: String) -> NSDate
+func convertPointAt(column column: Int, row: Int) -> CGPoint
+func timedAction(afterDelay delay: NSTimeInterval, perform action: SKAction) -> SKAction!
+
+// would be called like this:
+dateFromString("2014-03-14")
+convertPointAt(column: 42, row: 13)
+timedAction(afterDelay: 1.0, perform: someOtherAction)
 ```
 
-init() example:
+For methods, follow the standard Apple convention of referring to the first parameter in the method name:
+
 ```swift
-class Component {
-    var controller: CustomViewController! // We want to use it as a non-optional but have to initialize after super.init()
-    
-    init() {
-        // controller = CustomViewController(component: self) <-- compiler error
-        super.init()
-        controller = CustomViewController(component: self) 
-    }
+class Counter {
+  func combineWith(otherCounter: Counter, options: Dictionary?) { ... }
+  func incrementBy(amount: Int) { ... }
 }
 ```
 
-How do I handle errors in Swift?
-================================
+### Protocols
 
-As of Swift 2.0 exceptions have been introduced into the language and the APIs that formerly used `someMethod(..., error: *NSError)` make use of them. We have decided to stick with our existing pattern for returning errors.
-
-You have a couple options:
+Following Apple's API Design Guidelines, protocols names that describe what something is should be a noun. Examples: `Collection`, `WidgetFactory`. Protocols names that describe an ability should end in -ing, -able, or -ible. Examples: `Equatable`, `Resizing`.
 
-- Return a tuple (similar to Go)
-- Use a result/error enumeration type
+### Enumerations
 
-We typically use an enumeration type:
+Following Apple's API Design Guidelines for Swift 3, use lowerCamelCase for enumeration values.
 
 ```swift
-public enum ParseResult {
-    case Result(AST)
-    case Error(String)
+enum Shape {
+  case rectangle
+  case square
+  case rightTriangle
+  case equilateralTriangle
 }
 ```
 
+### Comments
+
+When referring to functions in comments include the required parameter names from the caller's perspective or `_` for unnamed parameters. Examples:
+
+> Call `convertPointAt(column:row:)` from your own `init` implementation.
+>
+> If you call `dateFromString(_:)` make sure that you provide a string with the format "yyyy-MM-dd".
+>
+> If you call `timedAction(afterDelay:perform:)` from `viewDidLoad()` remember to provide an adjusted delay value and an action to perform.
+>
+> You shouldn't call the data source method `tableView(_:cellForRowAtIndexPath:)` directly.
+
+This is the same as the `#selector` syntax. When in doubt, look at how Xcode lists the method in the jump bar – our style here matches that.
+
+### Class Prefixes
```

Comment:
I'm tempted to say we don't need this. I can't think of one time where we ran into this at Mobify. It is probably more of an issue for engineers coming from an Obj-C background where class prefixing was more of a "thing".